qtconsole/console_widget.py

```diff
@@ -998,18 +1008,40 @@ def _append_custom(self, insert, input, before_prompt=False, *args, **kwargs):
         current prompt, if there is one.
         """
         # Determine where to insert the content.
-        cursor = self._control.textCursor()
+        cursor = self._insert_text_cursor
         if before_prompt and (self._reading or not self._executing):
             self._flush_pending_stream()
-            cursor._insert_mode=True
-            cursor.setPosition(self._append_before_prompt_pos)
+
+            # Jump to before prompt, if there is one
+            if cursor.position() >= self._append_before_prompt_pos \
+                    and self._append_before_prompt_pos != self._get_end_pos():
+                cursor.setPosition(self._append_before_prompt_pos)
+
+                # If we appending on the same line as the prompt, use insert mode
+                # If so, the character at self._append_before_prompt_pos will not be a newline
+                cursor.movePosition(QtGui.QTextCursor.Right,
+                                    QtGui.QTextCursor.KeepAnchor)
+                if cursor.selection().toPlainText() != '\n':
+                    cursor._insert_mode = True
+                cursor.movePosition(QtGui.QTextCursor.Left)
         else:
+            # Insert at current printing point
+            # If cursor is before prompt jump to end, but only if there
+            # is a prompt (before_prompt_pos != end)
+            if cursor.position() <= self._append_before_prompt_pos \
+                    and self._append_before_prompt_pos != self._get_end_pos():
+                cursor.movePosition(QtGui.QTextCursor.End)
```

Comment:
A format suggestion:
```suggestion
            if (
                cursor.position() <= self._append_before_prompt_pos
                and self._append_before_prompt_pos != self._get_end_pos()
            ):
                cursor.movePosition(QtGui.QTextCursor.End)
```

Reply:
I do appreciate the code review. And I follow avoiding the \ when possible as a general guideline.
Though, it seems to be that the majority of multi-line if statements in this file use backslashes as opposed to explicitly encapsulated with parenthesis.

I'm fine with leaving this as is for now for consistency.

Reply:
Oh you are totally right! Makes sense to leave it with `\` for consistency :+1: